c/misra/src/rules/RULE-21-25/InvalidMemoryOrderArgument.ql

```diff
@@ -0,0 +1,172 @@
+    then ord = this.(MemoryOrderConstantAccess).getTarget()
+    else ord.getIntValue() = getValue().toInt()
+  }
+
+  /* Get the name of the `MemoryOrder` this expression is valued as. */
+  string getMemoryOrderString() { result = ord.toString() }
+}
+
+/**
+ * A `stdatomic.h` function which accepts a `memory_order` value as a parameter.
+ */
+class MemoryOrderedStdAtomicFunction extends Function {
+  int orderParamIdx;
+
+  MemoryOrderedStdAtomicFunction() {
+    exists(int baseParamIdx, int baseParams, string prefix, string suffix |
+      prefix = ["__", "__c11_"] and
+      suffix = ["", ".*", "_explicit"] and
+      (
+        getName().regexpMatch(prefix + ["atomic_thread_fence", "atomic_signal_fence"] + suffix) and
+        baseParamIdx = 0 and
+        baseParams = 1
+        or
+        getName()
+            .regexpMatch(prefix + ["atomic_load", "atomic_flag_clear", "atomic_flag_test_and_set"] +
+                suffix) and
+        baseParamIdx = 1 and
+        baseParams = 2
+        or
+        getName()
+            .regexpMatch(prefix + ["atomic_store", "atomic_fetch_.*", "atomic_exchange"] + suffix) and
+        baseParamIdx = 2 and
+        baseParams = 3
+        or
+        getName().regexpMatch(prefix + "atomic_compare_exchange_.*" + suffix) and
+        baseParamIdx = [3, 4] and
+        baseParams = 5
+      ) and
+      (
+        // GCC case, may have one or two inserted parameters, e.g.:
+        // __atomic_load(8, &repr->a, &desired, order)
+        // or
+        // __atomic_load_8(&repr->a, &desired, order)
+        prefix = "__" and
+        suffix = ".*" and
+        exists(int extraParams |
+          extraParams = getNumberOfParameters() - baseParams and
+          extraParams >= 0 and
+          orderParamIdx = baseParamIdx + extraParams
+        )
+        or
+        // Clang case, no inserted parameters:
+        // __c11_atomic_load(object, order)
+        suffix = "" and
+        prefix = "__c11_" and
+        orderParamIdx = baseParamIdx
+        or
+        // Non-macro case, may occur in a subset of gcc/clang functions:
+        prefix = "" and
+        suffix = "_explicit" and
+        orderParamIdx = baseParamIdx
+      )
+    )
+  }
+
+  int getOrderParameterIdx() { result = orderParamIdx }
+}
+
+module MemoryOrderFlowConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node node) {
+    // Direct usage of memory order constant
+    exists(MemoryOrderConstantAccess constant |
+      node.asExpr() = constant and
+      not constant.isAllowedOrder()
+    )
+    or
+    // A literal with a disallowed constant integer value
+    exists(Literal literal |
+      node.asExpr() = literal and
+      not literal.getValue().toInt() = any(AllowedMemoryOrder a).getValue().toInt()
+    )
+    or
+    // Everything else: not a memory order constant or an integer valued literal, also exclude
+    // variables and functions, things that flow further back.
```

Comment:
What else would this be? And if we cannot determine the value for it, presumably we can't tell whether it's in violation or not?

Reply:
The goal was to essentially look for proofs that the value is set correctly. All flow should track back to an explicitly correct value, something like a `BinaryOperation` should be incorrect, given that `memory_order_seq_cst` is a simple enum value.

I'm assuming, looking at this again, that this approach is inefficient. I deleted this case, but I opened https://github.com/github/codeql-coding-standards/issues/848 and simply deleted this case for now.

This gives us a false positive for `atomic_thread_fence(g1 + 1)`, which is reasonable.

Reply:
> This gives us a false positive for atomic_thread_fence(g1 + 1), which is reasonable.

I think it's a false negative, rather than a false positive!

> I'm assuming, looking at this again, that this approach is inefficient. I deleted this case, but I opened #848 and simply deleted this case for now.

I don't actually think it would be too inefficient - if I recall correctly, the shared dataflow library has some optimisations enabling the use of forward or reverse flow calculations to help with this case. I think it's fine to follow up in #848, though.